### PR TITLE
Fix deadlock in AtomicSetFlagAndRaiseSymbolDeclaredEvent

### PR DESCRIPTION
**Pull request by mavasani**
**Cause:** We are seeing intermittent deadlocks during build on linux machines running our test suites. The deadlock happens when one thread invokes AtomicSetFlagAndRaiseSymbolDeclaredEvent, and succeeds in setting the flag and invokes SymbolDeclaredEvent while holding onto the lock. The task scheduler decides to inline the event processing, which attempts to execute analyzers while holding onto the _diagnosticLock. Another thread meanwhile is already executing analyzers and calls back into the compiler and is waiting on _diagnosticLock.

**Fix:** Move the SymbolDeclaredEvent invocation outside the lock statement.

**Testing:** All tests pass locally.

**Fixes https://github.com/dotnet/roslyn/issues/6098**

### CODE REVIEW

**Comment by mavasani** [comment]
@AlekseyTs @gafter @JohnHamby can you please review?

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb [hunk 1]
```diff
@@ -887,19 +887,27 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
...
                         ' routine, which is a bug.
                         Throw ExceptionUtilities.Unreachable
                     End If
                 End If
-                Return change
             End SyncLock
+
+            If change Then
+                ' Invoke the symbol declared event outside the lock.
+                ' Task scheduler might decide to inline the event processing, which can call back into analyzers,
```
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb]: What does "inline" mean here?

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb [hunk 2]
```diff
@@ -887,19 +887,27 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
...
+                ' and we don't want that to happen while holding onto _diagnosticLock.
+                ' See https://github.com/dotnet/roslyn/issues/6098
```
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb]: The reference to the issue will become clutter almost immediately.

**Comment by JohnHamby** [comment]
I sign off, but you will need a signoff from someone who understands the compiler's eventing model better than I do.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb (hunk 1 above)
Comment (mavasani) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb]: The call stack shows that [SymbolDeclaredEvent](http://source.roslyn.io/Microsoft.CodeAnalysis.VisualBasic/R/d7a2310fa51fca81.html) invokes [EnqueuCore](http://source.roslyn.io/Microsoft.CodeAnalysis/R/329bc4b8e37cbd7f.html), which contains a  [SetResult](http://referencesource.microsoft.com/mscorlib/R/810e0ddcb0511ff1.html) invocation on a task completion source, which eventually ends up invoking [FinishContinuations](http://referencesource.microsoft.com/mscorlib/R/dadbf616e42976ee.html), which starts the worker task [ProcessCompilationEventsCoreAsync](http://source.roslyn.io/Microsoft.CodeAnalysis/R/39f4dbd78c629150.html) to process the event and execute analyzers.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb (hunk 2 above)
Comment (mavasani) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb]: I am hoping for our git repo to survive a lot longer!

**Comment by mavasani** [comment]
Closing this one as I mistakenly opened it against master branch. Opened https://github.com/dotnet/roslyn/pull/6172 against stabilization.

**Comment by mavasani** [comment]
Closing this one as I mistakenly opened it against master branch. Opened https://github.com/dotnet/roslyn/pull/6172 against stabilization.

**Comment by mavasani** [comment]
Closing this one as I mistakenly opened it against master branch. Opened https://github.com/dotnet/roslyn/pull/6172 against stabilization.

**Pull request closed without merging**